filecoin-proofs/src/bin/phase2.rs

```diff
@@ -355,492 +436,664 @@ fn create_initial_params<Tree: 'static + MerkleTreeTrait>(
+        "making streamer from small 'before' params: {}",
+        path_before
     );
 
-    info!("making contribution");
-    let start = Instant::now();
-    let contribution = params.contribute(&mut rng);
+    let mut streamer = if param_size.is_large() {
+        Streamer::new_from_large_file(path_before, read_raw, write_raw).unwrap_or_else(|e| {
+            panic!(
+                "failed to make streamer from large `{}`: {}",
+                path_before, e
+            );
+        })
+    } else {
+        Streamer::new(path_before, read_raw, write_raw).unwrap_or_else(|e| {
+            panic!(
+                "failed to make streamer from small `{}`: {}",
+                path_before, e
+            );
+        })
+    };
+
+    info!("writing small 'after' params to file: {}", path_after);
+    let file_after = File::create(&path_after).unwrap_or_else(|e| {
+        panic!(
+            "failed to create 'after' params file `{}`: {}",
+            path_after, e
+        );
+    });
+
+    let contrib = streamer
+        .contribute(&mut rng, file_after, CHUNK_SIZE)
+        .expect("failed to make contribution");
+
+    let contrib_str = hex_string(&contrib);
     info!(
-        "successfully made contribution, contribution hash: {}, dt_contribute={}s",
-        hex::encode(&contribution[..]),
-        start.elapsed().as_secs()
+        "successfully made contribution: {}, dt_contribute={}s",
+        contrib_str,
+        start_contrib.elapsed().as_secs()
     );
 
-    let path_after = params_filename(proof, hasher, sector_size, &head, param_number_before + 1);
-    info!("writing 'after' params to file: {}", path_after);
-    let file_after = File::create(path_after).unwrap();
-    let mut params_writer = BufWriter::with_capacity(1024 * 1024, file_after);
-    params.write(&mut params_writer).unwrap();
+    let contrib_path = format!("{}.contrib", path_after);
+    info!("writing contribution hash to file: {}", contrib_path);
+    fs::write(&contrib_path, contrib_str).unwrap_or_else(|e| {
+        panic!(
+            "failed to write contribution to file `{}`: {}",
+            contrib_path, e
+        );
+    });
+
     info!(
         "successfully made contribution, dt_total={}s",
         start_total.elapsed().as_secs()
     );
 }
 
-/// Verifies a sequence of parameter transitions against a sequence of corresponding contribution
-/// hashes. For example, verifies that the first digest in `contribution_hashes` transitions the
-/// first parameters file in `param_paths` to the second file, then verifies that the second
-/// contribution hash transitions the second parameters file to the third file.
-fn verify_param_transitions(param_paths: &[&str], contribution_hashes: &[[u8; 64]]) {
-    let start_total = Instant::now();
+fn convert_small<'a>(path_before: &'a str) {
+    let (proof, hasher, sector_size, head, param_number, param_size, read_raw) =
+        parse_params_filename(path_before);
```

Comment:
Could you add an assertion

```rust
# TODO: change this if we update the large MPC params (and G2Affine) to support the raw serialization format.
assert!(param_size.is_small(), "converting large params to raw format is not currently supported");
```

Reply:
And maybe the check

```rust
if read_raw {
    info!("params are already in the raw format, exiting early");
    return;
}
```

Reply:
I think your proposed assertion is already covered a few lines below:

```
    let mut streamer = if param_size.is_large() {
        panic!("cannot convert large param format");
```

If we *do* end up supporting raw format for large params, I assume we would not add it to the `convert_small` function anyway.

Reply:
Regarding the proposed check and early exit: this function actually converts both to and from raw. When called on params in raw format, it will convert to 'non-raw'.

Reply:
> [porcuquine]: I think your proposed assertion is already covered a few lines below
> [drpetervannostrant]: this assertion will remove a panic when streamer is created.
Yea, I just wanted to exit before the "converting params ..." log.

> this function actually converts both to and from raw
Sorry, I didn't realize that. Please disregard.